src/cookies.rs

```diff
@@ -1,46 +1,64 @@
 use cookie::{Cookie, CookieJar, ParseError};
 
 use crate::Context;
+use http::HeaderMap;
+use std::sync::{Arc, RwLock};
 
 /// A representation of cookies which wraps `CookieJar` from `cookie` crate
 ///
-/// Currently this only exposes getting cookie by name but future enhancements might allow more
-/// operations
-struct CookieData {
-    content: CookieJar,
+#[derive(Debug)]
+pub(crate) struct CookieData {
+    pub(crate) content: Arc<RwLock<CookieJar>>,
+}
+
+impl CookieData {
+    pub fn from_headers(headers: &HeaderMap) -> Self {
+        CookieData {
+            content: Arc::new(RwLock::new(
+                headers
+                    .get(http::header::COOKIE)
+                    .and_then(|raw| parse_from_header(raw.to_str().unwrap()).ok())
```

Comment:
I think we should terminate early and send an error response back here.

Reply:
Yes, mostly we should send an error response back early. I think `CookiesExt` methods should be changed to return a `Result`. Not sure we can send an error response(http response) from here.

Reply:
Yeah ideally, the extension should return a `Result` and the tide middleware handler should return an error response if any of the middleware functions return an error. I think that was the case previously, but it might have changed now. I'll have to check.